clippy_lints/src/methods/unnecessary_to_owned.rs

```diff
@@ -279,6 +280,21 @@ fn check_other_call_arg<'tcx>(
                 &trait_predicate.trait_ref.substs.iter().skip(1).collect::<Vec<_>>()[..],
                 call_substs,
             );
+
+            // if `expr` is string-like and generic target is [u8], skip
+            // (https://github.com/rust-lang/rust-clippy/issues/9317).
+            if let GenericArgKind::Type(arg_ty) = composed_substs[0].unpack() {
+                if arg_ty.is_slice() {
+                    let inner_ty = arg_ty.builtin_index().unwrap();
+                    if matches!(inner_ty.kind(), ty::Uint(ty::UintTy::U8)) {
+                        let self_ty = cx.typeck_results().expr_ty(expr).peel_refs();
+                        if  is_type_diagnostic_item(cx, self_ty, sym::String) {
+                            return false;
+                        }
+                    }
+                }
+            }
+
             implements_trait(cx, receiver_ty, as_ref_trait_id, &composed_substs)
```

Comment:
```suggestion
            // if `expr` is a `String` and generic target is [u8], skip
            // (https://github.com/rust-lang/rust-clippy/issues/9317).
            if let [subst] = composed_substs
                && let GenericArgKind::Type(arg_ty) = subst.unpack() 
                && arg_ty.is_slice() 
                && let inner_ty = arg_ty.builtin_index().unwrap()
                && let ty::Uint(ty::UintTy::U8) = inner_ty.kind() 
                && let self_ty = cx.typeck_results().expr_ty(expr).peel_refs()
                && is_type_diagnostic_item(cx, self_ty, sym::String) {
                false
            } else {
                implements_trait(cx, receiver_ty, as_ref_trait_id, &composed_substs)
            }
```

Let's use `let_chain`s here.

Reply:
https://github.com/rust-lang/rust-clippy/compare/3fa0fba225454cb3aebf5dd973827d6f540de5cf..1a2aaf68de4989eaacf5b669da00b50dca8b41e0